kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+    if (ret == 0)
+        ret = kadm5_get_principal(r->kadm_handle, p, &princ, mask);
+    if (ret == 0) {
+        freeit = 1;
+
+        /*
+         * If princ is virtual and we're not asked to materialize, ignore
+         * requests to rotate.
+         */
+        if (!r->materialize &&
+            (princ.attributes & (KRB5_KDB_VIRTUAL_KEYS | KRB5_KDB_VIRTUAL))) {
+            r->rotate = 0;
+            r->revoke = 0;
+        }
+    }
+
+    change = !r->ro && (r->rotate || r->revoke);
+
+    /* Handle create / materialize options */
+    if (ret == KADM5_UNK_PRINC && r->create) {
+        char pw[128];
+
+        ret = strcmp(r->method, "POST") == 0 ? 0 : ENOSYS; /* XXX */
+
+        /*
+         * We're writing, but maybe we can't (not a primary) or have to
+         * reconnect to a primary kadmind.
+         */
+        if (ret == 0 && primary_server && !primary_admin_server) {
+            /* Local DB, not a primary -> redirect */
+            ret = KADM5_READ_ONLY;
+        } else if (ret == 0 && primary_admin_server) {
+            /* Connected to replica -> reconnect to primary */
+            kadm5_destroy(r->kadm_handle);
+            ret = get_kadm_handle(r->context, &r->kadm_handle, 1);
+        } else if (ret == 0 && local_read_only) {
+            ret = KADM5_READ_ONLY;
+        }
+        memset(&princ, 0, sizeof(princ));
+        princ.kvno = 1;
+        princ.tl_data = NULL;
+        princ.key_data = NULL;
+        princ.max_life = 24 * 3600;                /* XXX Make configurable */
+        princ.max_renewable_life = princ.max_life; /* XXX Make configurable */
+
+        random_password(r->context, pw, sizeof(pw));
+        princ.principal = p;     /* Borrow */
+        if (ret == 0)
+            ret = kadm5_create_principal(r->kadm_handle, &princ, create_mask,
+                                         pw);
+        princ.principal = NULL;  /* Return */
+        refetch = 1;
+        freeit = 1;
+    } else if (ret == 0 && r->materialize &&
+               (princ.attributes & KRB5_KDB_VIRTUAL)) {
+
+        ret = strcmp(r->method, "POST") == 0 ? 0 : ENOSYS; /* XXX */
+
+        /*
+         * We're writing, but maybe we can't (not a primary) or have to
+         * reconnect to a primary kadmind.
+         */
+        if (ret == 0 && primary_server && !primary_admin_server) {
+            /* Local DB, not a primary -> redirect */
+            ret = KADM5_READ_ONLY;
+        } else if (ret == 0 && primary_admin_server) {
+            /* Connected to replica -> reconnect to primary */
+            kadm5_destroy(r->kadm_handle);
+            ret = get_kadm_handle(r->context, &r->kadm_handle, 1);
+        } else if (ret == 0 && local_read_only) {
+            ret = KADM5_READ_ONLY;
+        }
+        princ.attributes |= KRB5_KDB_MATERIALIZE;
+        princ.attributes &= ~KRB5_KDB_VIRTUAL;
+        /*
+         * XXX If there are TL data which should be re-encoded and sent as
```

Comment:
This needs to be done.  So either disable materialization for now, or do what the comment says.